### PR TITLE
refactor(core): Use type-safe event emitters (no-changelog)

### PR DESCRIPTION
**Pull request by netroy**
## Summary
This PR updates all EventEmitters in `cli` (except `MessageEventBus`) to be type-safe.

## Review / Merge checklist

- [x] PR title and summary are descriptive

### CODE REVIEW

File: packages/cli/src/TypedEmitter.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+type EventName = string;
+
+type Payloads<L> = {
+	[E in Extract<keyof L, EventName>]: unknown;
+};
+
+type Listener<P> = (payload: P) => void;
+
+export class TypedEmitter<L extends Payloads<L>> extends EventEmitter {
+	protected debounceWait = 300;
```
Comment (ivov) [inline packages/cli/src/TypedEmitter.ts:13]:
```suggestion
	protected debounceWait = 300; // ms
```

File: packages/cli/src/concurrency/concurrency-queue.ts [hunk 2]
```diff
@@ -1,9 +1,14 @@
 import { Service } from 'typedi';
-import { EventEmitter } from 'node:events';
-import debounce from 'lodash/debounce';
+import { TypedEmitter } from '@/TypedEmitter';
+
+interface Events {
```
Comment (ivov) [inline packages/cli/src/concurrency/concurrency-queue.ts:4]: Interfaces are subject to declaration merging so if we ever export we can accidentally create a big `Events` interface. Let's use `type` instead? Same elsewhere.

Comment (ivov) [reply packages/cli/src/concurrency/concurrency-queue.ts:4]: Also for clarity and easier searching I'd rename based on the service, e.g. `ConcurrencyEvents`

File: packages/cli/src/services/workflow-statistics.service.ts [hunk 3]
```diff
@@ -90,7 +109,7 @@ export class WorkflowStatisticsService extends EventEmitter {
 		const owner = await this.ownershipService.getProjectOwnerCached(project.id);
 
 		let metrics = {
-			user_id: owner?.id,
+			user_id: owner!.id,
```
Comment (ivov) [inline packages/cli/src/services/workflow-statistics.service.ts:112]: No context here but `owner` looks like a project, not a user.

File: packages/cli/src/TypedEmitter.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+	override on<U extends Extract<keyof L, EventName>>(event: U, listener: Listener<L[U]>) {
```
Comment (ivov) [inline packages/cli/src/TypedEmitter.ts:15]: Why `U` and not `E`? In general I find all the abbreviations and `Extract` uses might be making this more opaque that it needs to be, esp. as we know in an `EventEmitter` all keys are strings anyway. Here's my take if you think it's useful:

```ts
import { EventEmitter } from 'node:events';
import debounce from 'lodash/debounce';

type Payloads<ListenerMap> = {
	[E in keyof ListenerMap]: unknown;
};

type Listener<Payload> = (payload: Payload) => void;

export class TypedEmitter<ListenerMap extends Payloads<ListenerMap>> extends EventEmitter {
	private debounceWait = 300;

	override on<EventName extends keyof ListenerMap & string>(
		eventName: EventName,
		listener: Listener<ListenerMap[EventName]>,
	) {
		return super.on(eventName, listener);
	}

	override once<EventName extends keyof ListenerMap & string>(
		eventName: EventName,
		listener: Listener<ListenerMap[EventName]>,
	) {
		return super.once(eventName, listener);
	}

	override off<EventName extends keyof ListenerMap & string>(
		eventName: EventName,
		listener: Listener<ListenerMap[EventName]>,
	) {
		return super.off(eventName, listener);
	}

	override emit<EventName extends keyof ListenerMap & string>(
		eventName: EventName,
		payload?: ListenerMap[EventName],
	): boolean {
		return super.emit(eventName, payload);
	}

	protected debouncedEmit = debounce(
		<EventName extends keyof ListenerMap & string>(
			eventName: EventName,
			payload?: ListenerMap[EventName],
		) => super.emit(eventName, payload),
		this.debounceWait,
	);
}
```

**Review by ivov — COMMENTED** [review]
Love this!

File: packages/cli/src/services/workflow-statistics.service.ts (hunk 3 above)
Comment (ivov) [reply packages/cli/src/services/workflow-statistics.service.ts:112]: @valya Is this a bug?

**Review by ivov — APPROVED** [review]
Now all we need is a lint rule forbidding `extends EventEmitter` except on `TypedEmitter` :)

File: packages/cli/src/services/workflow-statistics.service.ts (hunk 3 above)
Comment (netroy) [reply packages/cli/src/services/workflow-statistics.service.ts:112]: `owner` is a `User` 
image

**Comment by janober** [comment]
Got released with [`n8n@1.53.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.53.0)

**Pull request merged**